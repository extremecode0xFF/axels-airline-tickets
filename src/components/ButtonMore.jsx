import React from 'react';
import { Button, Container } from 'react-bootstrap';
import styled from 'styled-components';

const Styles = styled(Container)`
  .btn {
    transition: all 0.2s ease-out;
    line-height: 50px;
    background-color: #2196f3;
    :focus {
      outline: none;
      box-shadow: none;
    }
    :hover {
      background-color: #0186f3;
      border: 1px solid black;
    }
  }
`;

export default function ButtonMore() {
  return (
    <Styles>
      <Button className="w-100" size="lg">
        Показать еще пять билетов!
      </Button>
    </Styles>
  );
}
