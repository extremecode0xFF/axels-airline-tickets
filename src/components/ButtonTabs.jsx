import React from 'react';

import { Button, ButtonGroup } from 'react-bootstrap';
import { ContainerButtonTabs } from '../styled/ButtonTabs';

import { useSearchParams } from 'react-router-dom';

const ButtonTabs = ({ config }) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const handleClick = (event) => {
    if (!event.target.value) return;
    const [key, value] = event.target.value.split(',');
    if (searchParams.has(key)) return;

    config.forEach((val) => {
      const [param] = Object.keys(val.query);
      if (searchParams.has(param)) {
        searchParams.delete(param);
      }
    });

    searchParams.append(key, value);

    setSearchParams({
      ...Object.fromEntries([...searchParams]),
    });
  };

  return (
    <ContainerButtonTabs className="mb-3">
      <ButtonGroup className="w-100" size="dm" onClick={handleClick}>
        {config.map((filter, index) => (
          <Button
            className="text-uppercase"
            variant="light"
            key={index}
            id={filter.title}
            value={Object.entries(filter.query).join()}
            disabled={searchParams.has(Object.keys(filter.query))}
          >
            {filter.title}
          </Button>
        ))}
      </ButtonGroup>
    </ContainerButtonTabs>
  );
};

export default ButtonTabs;
