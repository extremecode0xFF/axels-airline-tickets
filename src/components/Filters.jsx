import React from 'react';

import { Container, Form } from 'react-bootstrap';
import { ContainerFilters } from '../styled/Filters';

import { useSearchParams } from 'react-router-dom';

const Filters = ({ name, config }) => {
  const [searchParams, setSearchParams] = useSearchParams();

  const handleChange = (event) => {
    const [key, value] = event.target.value.split(',');

    if (searchParams.has(key)) {
      searchParams.delete(key);
    } else {
      searchParams.append(key, value);
    }

    setSearchParams(searchParams);
  };

  return (
    <ContainerFilters className="mb-3 shadow-sm">
      <Form>
        <p className="text-uppercase mx-3 my-0 pt-4 pb-2">{name}</p>
        {config.map((param, i) => {
          return (
            <Container key={i} className="content p-0 py-2">
              <Form.Check
                className="d-flex justify-content-between align-items-center mx-3 user-select-none"
                type="checkbox"
                id={Object.keys(param.query)}
                label={param.title}
                value={Object.entries(param.query).join()}
                onChange={handleChange}
                checked={searchParams.has(Object.keys(param.query))}
              />
            </Container>
          );
        })}
      </Form>
    </ContainerFilters>
  );
};

export default Filters;
