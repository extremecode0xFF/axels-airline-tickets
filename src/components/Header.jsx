import React from 'react';
import logo from '../assets/images/logo.svg';
import { Image, Container } from 'react-bootstrap';

export default function Header() {
  return (
    <Container className="d-flex justify-content-center py-5">
      <Image src={logo} alt="logo" />
    </Container>
  );
}
