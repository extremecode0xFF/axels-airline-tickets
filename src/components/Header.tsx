import React from 'react';

import { Image, Container } from 'react-bootstrap';

const Header: React.FC = () => (
  <Container className="d-flex justify-content-center py-5">
    <Image src="images/header_logo.svg" alt="logo" />
  </Container>
);

export default Header;
