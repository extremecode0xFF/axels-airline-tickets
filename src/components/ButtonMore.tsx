import React, { FC } from 'react';

import { Button } from 'react-bootstrap';
import { ContainerButtonMore } from '../styled/ButtonMore';

type PropTypes = {
  onClick: () => void;
};

const ButtonMore: FC<PropTypes> = ({ onClick }) => (
  <ContainerButtonMore>
    <Button className="w-100 mb-3" size="lg" onClick={onClick}>
      Показать еще 5 билетов!
    </Button>
  </ContainerButtonMore>
);

export default ButtonMore;
