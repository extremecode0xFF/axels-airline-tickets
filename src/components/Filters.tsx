import React, { useEffect, useMemo, memo, ChangeEvent } from 'react';
import { useSearchParams } from 'react-router-dom';
import { useFormik } from 'formik';
import { useDispatch } from 'react-redux';

import { Container, Form } from 'react-bootstrap';
import { ContainerFilters } from '../styled/Filters';

import { setFilteredTickets } from '../redux/ducks/tickets';
import { ConfigParams } from '../types/config';

interface PropTypes {
  name: string;
  config: ConfigParams[];
}

const Filters = memo<PropTypes>(({ name, config }) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const dispatch = useDispatch();

  const initParams = useMemo(
    () =>
      config.reduce<Record<string, boolean>>((acc, { query }) => {
        if (searchParams.has(query)) {
          acc[query] = true;
        } else {
          acc[query] = false;
        }
        return acc;
      }, {}),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const formik = useFormik({
    initialValues: {
      checkAll: false,
      checked: initParams,
    },
    onSubmit() {},
  });

  useEffect(() => {
    const params = Object.keys(formik.values.checked).reduce<string[]>(
      (acc, param) => {
        let isChecked = formik.values.checked[param];
        if (isChecked) {
          acc.push(param);
        }
        return acc;
      },
      []
    );
    dispatch(setFilteredTickets(params));
  }, [dispatch, formik.values.checked]);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const { id, checked } = event.currentTarget;
    if (checked) {
      searchParams.set(id, '');
    } else {
      searchParams.delete(id);
    }
    setSearchParams(searchParams);
  };

  const handleChangeAll = (event: ChangeEvent<HTMLInputElement>) => {
    const { checked } = event.currentTarget;
    const newCheckedField: Record<string, boolean> = {};
    for (const param in formik.values.checked) {
      if (checked) {
        searchParams.set(param, '');
        newCheckedField[param] = true;
      } else {
        searchParams.delete(param);
        newCheckedField[param] = false;
      }
    }
    formik.setFieldValue(`checked`, newCheckedField);
    setSearchParams(searchParams);
  };

  return (
    <ContainerFilters className="mb-3 shadow-sm">
      <p className="text-uppercase mx-3 my-0 pt-4 pb-2">{name}</p>
      <Container className="content p-0 py-2">
        <Form.Check
          className="d-flex justify-content-between align-items-center mx-3 user-select-none"
          label="Все"
          id="checkAll"
          onChange={(event) => {
            formik.handleChange(event);
            handleChangeAll(event);
          }}
          checked={formik.values.checkAll}
        />
      </Container>
      {config.map(({ title, query }) => {
        return (
          <Container key={query} className="content p-0 py-2">
            <Form.Check
              className="d-flex justify-content-between align-items-center mx-3 user-select-none"
              name={`checked[${query}]`}
              id={query}
              label={title}
              onChange={(event) => {
                formik.handleChange(event);
                handleChange(event);
              }}
              checked={formik.values.checked[query]}
            />
          </Container>
        );
      })}
    </ContainerFilters>
  );
});

export default Filters;
