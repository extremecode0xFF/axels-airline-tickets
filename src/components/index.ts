import ButtonMore from './ButtonMore';
import ButtonTabs from './ButtonTabs';
import Filters from './Filters';
import Header from './Header';
import Ticket from './Ticket';
import TicketList from './TicketList';

export { ButtonMore, ButtonTabs, Filters, Header, Ticket, TicketList };
