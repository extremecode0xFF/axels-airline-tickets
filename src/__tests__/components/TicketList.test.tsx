import React from 'react';
import renderer from 'react-test-renderer';

import { TicketList } from '../../components';

import FAKE_TICKETS from '../../data/constants/fakeTickets';

describe('TicketList', () => {
  const tickets = FAKE_TICKETS;
  test('it should match a snapshot', () => {
    const tree = renderer
      .create(<TicketList tickets={tickets} visibleTickets={5} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
