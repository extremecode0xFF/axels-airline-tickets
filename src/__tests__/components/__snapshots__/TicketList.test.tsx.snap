// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketList match snapshot 1`] = `
<div
  className="container"
>
  <div
    className="sc-hKMtZM dtidkA mb-3 p-4 bg-white rounded vstack gap-3"
  >
    <div
      className="d-flex justify-content-between align-items-center container"
    >
      <span
        className="price"
      >
        20 210 P
      </span>
      <img
        alt="airline_log"
        className=""
        src="airlines_logo.png"
      />
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          AA-BB
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          1ч 40м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          0 пересадок
        </span>
        <span
          className="info"
        />
      </div>
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          AA-BB
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          0ч 50м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          0 пересадок
        </span>
        <span
          className="info"
        />
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM dtidkA mb-3 p-4 bg-white rounded vstack gap-3"
  >
    <div
      className="d-flex justify-content-between align-items-center container"
    >
      <span
        className="price"
      >
        22 000 P
      </span>
      <img
        alt="airline_log"
        className=""
        src="airlines_logo.png"
      />
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          AA-BB
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          2ч 5м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          1 пересадка
        </span>
        <span
          className="info"
        >
          CC
        </span>
      </div>
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          AA-BB
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          2ч 50м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          1 пересадка
        </span>
        <span
          className="info"
        >
          CC
        </span>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM dtidkA mb-3 p-4 bg-white rounded vstack gap-3"
  >
    <div
      className="d-flex justify-content-between align-items-center container"
    >
      <span
        className="price"
      >
        19 000 P
      </span>
      <img
        alt="airline_log"
        className=""
        src="airlines_logo.png"
      />
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          AA-BB
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          0ч 20м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          2 пересадки
        </span>
        <span
          className="info"
        >
          AB,BC
        </span>
      </div>
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          CV-VC
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          0ч 10м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          2 пересадки
        </span>
        <span
          className="info"
        >
          GB,BG
        </span>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM dtidkA mb-3 p-4 bg-white rounded vstack gap-3"
  >
    <div
      className="d-flex justify-content-between align-items-center container"
    >
      <span
        className="price"
      >
        23 400 P
      </span>
      <img
        alt="airline_log"
        className=""
        src="airlines_logo.png"
      />
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          MS-VC
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          0ч 10м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          3 пересадки
        </span>
        <span
          className="info"
        >
          AB,BC,CC
        </span>
      </div>
    </div>
    <div
      className="d-flex justify-content-between container"
    >
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title"
        >
          CV-VC
        </span>
        <span
          className="info"
        >
          NaN:NaN - NaN:NaN
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          В пути
        </span>
        <span
          className="info"
        >
          0ч 10м
        </span>
      </div>
      <div
        className="d-flex flex-column p-0 container"
      >
        <span
          className="title text-uppercase"
        >
          3 пересадки
        </span>
        <span
          className="info"
        >
          GB,BG,CC
        </span>
      </div>
    </div>
  </div>
</div>
`;
